### PR TITLE
Persistent notification, added 2 parallel uploads and updated work policies

### PR DESCRIPTION
**Pull request by zafirskthelifehacker**
- Added a Restriction(Used Semaphore) for letting Work Manager know that is only 2 threads for 2 Parallel Uploads
- Set Notification to No Clear(Added more Flags) - Makes Notifs more persistent
- Updated Background Policy to Linear
- Tested Work Policy on both APPEND and KEEP(Use KEEP)

### CODE REVIEW

File: src/android/FileTransferBackground.java [hunk 1]
```diff
@@ -1,13 +1,22 @@
 package com.spoon.backgroundfileupload;
 
+import static android.app.NotificationManager.INTERRUPTION_FILTER_NONE;
```
Comment (dinitri-ragoo) [inline src/android/FileTransferBackground.java:3]:
```suggestion
```

File: src/android/FileTransferBackground.java [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
+
 import android.app.NotificationChannel;
 import android.app.NotificationManager;
 import android.content.Context;
+import android.net.ConnectivityManager;
+import android.net.Network;
+import android.net.NetworkCapabilities;
+import android.net.NetworkInfo;
+import android.net.NetworkRequest;
```
Comment (dinitri-ragoo) [inline src/android/FileTransferBackground.java:12]:
```suggestion
```

File: src/android/FileTransferBackground.java [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
 import android.os.Build;
 import android.util.Log;
 
 import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
```
Comment (dinitri-ragoo) [inline src/android/FileTransferBackground.java:17]:
```suggestion
```

File: src/android/FileTransferBackground.java [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 import androidx.work.BackoffPolicy;
+import androidx.work.Configuration;
```
Comment (dinitri-ragoo) [inline src/android/FileTransferBackground.java:19]:
```suggestion
```

File: src/android/FileTransferBackground.java [hunk 5]
```diff
@@ -37,6 +46,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
```
Comment (dinitri-ragoo) [inline src/android/FileTransferBackground.java:49]:
```suggestion
```

**Review by dinitri-ragoo — COMMENTED** [review]
remove unused imports + configure concurrent uploads from js to java

**Comment by zafirskthelifehacker** [comment]
**Notification Messages**
- Some image(s) is left to be uploaded
- Some image(s) is left to be uploaded(Problem with network)
- Some image(s) has not been uploaded
- Some image(s) has not been uploaded(Due to some network issues)

File: src/android/UploadTask.java [hunk 6]
```diff
@@ -207,6 +244,24 @@ public Result doWork() {
...
+                }
+            } else {
+                Log.d(TAG, "WIFI is not connected");
+                setForegroundAsync(UploadForegroundNotification.getRetryForegroundInfo(getApplicationContext()));
+            }
+        } else {
+            Log.d(TAG, "WIFI is not connected");
+            setForegroundAsync(UploadForegroundNotification.getRetryForegroundInfo(getApplicationContext()));
+        }
+
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:264]: we can simplify this

File: src/android/UploadTask.java [hunk 7]
```diff
@@ -323,6 +388,24 @@ private void handleProgress(long bytesWritten, long totalBytes) {
...
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:408]: duplicate piece of code

File: src/android/WorkManagerInit.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    public int delete(@NonNull Uri uri, @Nullable String s, @Nullable String[] strings) {
+        return 0;
+    }
+
+    @Override
+    public int update(@NonNull Uri uri, @Nullable ContentValues contentValues, @Nullable String s, @Nullable String[] strings) {
+        return 0;
+    }
+}
```
Comment (dinitri-ragoo) [inline src/android/WorkManagerInit.java:56]: missing empty line after last line of code

**Review by dinitri-ragoo — CHANGES_REQUESTED** [review]

**Comment by dinitri-ragoo** [comment]
Notification Suggestion

**Upload interrupted**
Please check your internet connection or try to connect to another network to see if your upload resumes.

File: README.md [hunk 9]
```diff
@@ -32,6 +32,15 @@ var config = {};
 var uploader = FileTransferManager.init(config, callback);
 ```
 
+**Setting up Work Manager for Parallel Upload**
```
Comment (dinitri-ragoo) [inline README.md:35]:
```suggestion
**Setting up Work Manager for concurrent uploads**
```

**Review by dinitri-ragoo — CHANGES_REQUESTED** [review]
- maybe it's time to change parallel upload to concurrent upload
- unnecessary comments can be removed
- does this offline notification shows up while there is no upload in progress ?

File: src/android/UploadTask.java [hunk 10]
```diff
@@ -230,6 +272,9 @@ public Result doWork() {
                     .putBoolean(KEY_OUTPUT_FAILURE_CANCELED, false)
                     .build()
             );
+        } catch (NullPointerException e) {
+//            setForegroundAsync(UploadForegroundNotification.getRetryForegroundInfo(getApplicationContext()));
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:276]:
```suggestion
```
pourquoi on garde ca ?

File: src/android/UploadTask.java [hunk 11]
```diff
@@ -359,6 +411,12 @@ private Request createRequest() throws FileNotFoundException {
         File file = new File(filepath);
         ProgressRequestBody fileRequestBody = new ProgressRequestBody(mediaType, file.length(), new FileInputStream(file), this::handleProgress);
 
+        //Create a BroadcastReceiver to check status of internet connectivity
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:414]:
```suggestion
        // Create a BroadcastReceiver to check status of internet connectivity
```

File: src/android/UploadTask.java [hunk 12]
```diff
@@ -457,4 +515,15 @@ public void writeTo(@NonNull BufferedSink bufferedSink) throws IOException {
             }
         }
     }
+
+    private class NetworkReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+            if((connectivityManager == null) || (connectivityManager.getActiveNetworkInfo() == null) || (connectivityManager.getActiveNetworkInfo().isConnected() == false)) {
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:523]: we can negate the "truthy" condition to make the line shorter

File: src/android/UploadTask.java [hunk 13]
```diff
@@ -457,4 +515,15 @@ public void writeTo(@NonNull BufferedSink bufferedSink) throws IOException {
...
+                Log.d(TAG, "WIFI is not connected");
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:524]: does this network check only cater for WIFI connection ? does it take into account mobile data ?

**Review by dinitri-ragoo — CHANGES_REQUESTED** [review]

File: src/android/UploadTask.java (hunk 12 above)
Comment (zafirskthelifehacker) [reply src/android/UploadTask.java:523]: okay will test that

File: src/android/UploadTask.java (hunk 13 above)
Comment (zafirskthelifehacker) [reply src/android/UploadTask.java:524]: Yes it check for any network connection

Comment (dinitri-ragoo) [reply src/android/UploadTask.java:524]: so a better log message could be  "No internet connection" ? @zafirskthelifehacker

Comment (zafirskthelifehacker) [reply src/android/UploadTask.java:524]: @dinitri-ragoo yeps

File: src/android/UploadTask.java (hunk 10 above)
Comment (zafirskthelifehacker) [reply src/android/UploadTask.java]: @dinitri-ragoo removed

**Comment by zafirskthelifehacker** [comment]
@dinitri-ragoo please review

File: src/android/UploadTask.java [hunk 14]
```diff
@@ -160,23 +176,68 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
+            Notification retryNotification = new NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
+                    .setContentTitle(notificationRetryTitle)
+                    .setTicker(notificationRetryTitle)
+                    .setContentText(notificationRetryText)
+                    .setSmallIcon(notificationIconRes)
+                    .setColor(Color.rgb(57, 100, 150))
+                    .build();
+
+            retryCachedInfo = new ForegroundInfo(notificationId, retryNotification);
+            return retryCachedInfo;
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:207]:
```suggestion
            return new ForegroundInfo(notificationId, retryNotification);
```

don't forget to also remove retryCachedInfo from attributes

File: src/android/UploadTask.java [hunk 15]
```diff
@@ -160,23 +176,68 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
                     .setTicker(notificationTitle)
                     .setSmallIcon(notificationIconRes)
                     .setColor(Color.rgb(57, 100, 150))
+                    // Notify device that the notification is for an ongoing process
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:179]:
```suggestion
```
remove comment

File: src/android/UploadTask.java [hunk 16]
```diff
@@ -160,23 +176,68 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
                     .setProgress(100, (int) (totalProgress * 100f), false)
                     .build();
 
+            // Not allow user to clear
+            notification.flags |= Notification.FLAG_NO_CLEAR;
+            // Notify device that the notification is for an ongoing process
+            // Some device support this flag
+            notification.flags |= Notification.FLAG_ONGOING_EVENT;
+            // Notify device that the notification is for a currently running foreground service
+            notification.flags |= Notification.FLAG_FOREGROUND_SERVICE;
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:190]:
```suggestion
            notification.flags |= Notification.FLAG_NO_CLEAR;
            notification.flags |= Notification.FLAG_ONGOING_EVENT;
            notification.flags |= Notification.FLAG_FOREGROUND_SERVICE;
```
remove comments

File: src/android/UploadTask.java [hunk 17]
```diff
@@ -160,23 +176,68 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
 
     private static OkHttpClient httpClient;
 
     private Call currentCall;
 
+    private static int concurency = 1;
+
+    private static Semaphore concurrentUploads = new Semaphore(concurency, true);
+
+    private static Semaphore concurrencyLock = new Semaphore(1);
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:226]:
```suggestion
    private static int concurency = 1;
    private static Semaphore concurrentUploads = new Semaphore(concurency, true);
    private static Semaphore concurrencyLock = new Semaphore(1);
```
we can group related attributes, for example here, all 3 are related to concurrency

**Review by dinitri-ragoo — COMMENTED** [review]
remove package-lock

File: src/android/UploadTask.java [hunk 18]
```diff
@@ -92,15 +105,17 @@
...
-    private static class UploadForegroundNotification {
+    public static class UploadForegroundNotification extends Service {
         private static final Map<UUID, Float> collectiveProgress = Collections.synchronizedMap(new HashMap<>());
         private static final AtomicLong lastNotificationUpdateMs = new AtomicLong(0);
         private static ForegroundInfo cachedInfo;
 
         private static final int notificationId = new Random().nextInt();
-        private static String notificationTitle = "Default title";
+        public static String notificationTitle = "Default title";
+        public static String notificationRetryTitle = "Upload interrupted";
```
Comment (dinitri-ragoo) [inline src/android/UploadTask.java:115]:
```suggestion
        public static String notificationRetryTitle = "Upload paused";
```

File: src/android/UploadTask.java [hunk 19]
```diff
@@ -240,10 +294,24 @@ public Result doWork() {
...
+                try {
+                    try {
+                        concurrentUploads.acquire();
+                        try {
+                            response = currentCall.execute();
+                        } finally {
+                            concurrentUploads.release();
+                        }
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
```
Comment (ombr) [inline src/android/UploadTask.java:309]: @zafirskthelifehacker here I think we might have an issue we are not telling the plug-in to do a retry if it's has been interrupted

Comment (zafirskthelifehacker) [reply src/android/UploadTask.java:309]: @ombr Yeah there should be a retry when it has been interrupted. I have added return Result.retry() for this catch statement.

**Review by dinitri-ragoo — COMMENTED** [review]
can we handle the onclick notif on the cordova app itself instead of the plugin ? Here we have no control over where to redirect user on notification click

File: src/android/UploadTask.java [hunk 20]
```diff
@@ -145,38 +173,116 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
 
+    private static int concurency = 1;
+    private static Semaphore concurrentUploads = new Semaphore(concurency, true);
+    private static Semaphore concurrencyLock = new Semaphore(1);
+
     public UploadTask(@NonNull Context context, @NonNull WorkerParameters workerParams) {
+
         super(context, workerParams);
 
+        int concurrencyConfig = workerParams.getInputData().getInt(KEY_INPUT_CONFIG_CONCURRENT_DOWNLOADS, 1);
```
Comment (azharbeebeejaun) [inline src/android/UploadTask.java:277]: On s'était dit de mettre à 2 par défaut, non ?

File: src/android/UploadTask.java [hunk 21]
```diff
@@ -145,38 +173,116 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
     }
     // </editor-fold>
 
     private static OkHttpClient httpClient;
 
     private Call currentCall;
 
+    private static int concurency = 1;
+    private static Semaphore concurrentUploads = new Semaphore(concurency, true);
+    private static Semaphore concurrencyLock = new Semaphore(1);
```
[the remark this answers is not in the document]
Comment (ombr) [reply src/android/UploadTask.java]: C'est pas tout à fait la même chose mais un sémaphore permet de gérer la concurrence.

File: src/android/UploadTask.java [hunk 22]
```diff
@@ -144,39 +183,150 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
+    private static Mutex concurrencyLock = new Mutex();
+
     public UploadTask(@NonNull Context context, @NonNull WorkerParameters workerParams) {
+
         super(context, workerParams);
 
+        int concurrencyConfig = workerParams.getInputData().getInt(KEY_INPUT_CONFIG_CONCURRENT_DOWNLOADS, 1);
+
+        try {
+            concurrencyLock.acquire();
```
Comment (ombr) [inline src/android/UploadTask.java:320]: @azharbeebeejaun @zafirskthelifehacker il faut aussi que l'on parle de ça ;-)

**Review by ombr — COMMENTED** [review]
Il faut que l on parle du mutex (il est pas forcément utile)

File: src/android/UploadTask.java (hunk 22 above)
Comment (azharbeebeejaun) [reply src/android/UploadTask.java:329]: @ombr il y a ça aussi

File: src/android/UploadTask.java [hunk 23]
```diff
@@ -144,39 +183,150 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
+            if (concurrency != concurrencyConfig) {
+                concurrency = concurrencyConfig;
+                concurrentUploads = new Semaphore(concurrencyConfig, true);
+            }
+            concurrencyLock.release();
```
Comment (ombr) [inline src/android/UploadTask.java:334]:
```suggestion
            concurrencyLock.acquire();
            try {
              if (concurrency != concurrencyConfig) {
                  concurrency = concurrencyConfig;
                  concurrentUploads = new Semaphore(concurrencyConfig, true);
              }
            } finally {
              concurrencyLock.release();
           }
```

**Review by dinitri-ragoo — CHANGES_REQUESTED** [review]

File: src/android/UploadTask.java [hunk 24]
```diff
@@ -144,39 +198,135 @@ private static ForegroundInfo getForegroundInfo(final Context context) {
...
 
-            totalProgress /= (float) uploadCount;
+            if (uploadCount > previousUploadCount) {
+                totalProgressStore += totalProgress / uploadCount;
+                previousUploadCount = uploadCount + (uploadCount - previousUploadCount);
+            } else if (uploadCount == previousUploadCount) {
+                totalProgressStore += totalProgress / previousUploadCount;
+            } else {
+                totalProgressStore += totalProgress / (previousUploadCount - uploadCount);
+            }
```
[the remark this answers is not in the document]
Comment (azharbeebeejaun) [reply src/android/UploadTask.java:215]: Ok pour moi !

**Pull request merged**